include/sparrow/builder/builder.hpp

```diff
@@ -0,0 +1,406 @@
+        });
+
+        return type(
+            std::move(detyped_children),
+            where_null(t)
+       );
+    }
+};
+
+template< translate_to_variable_sized_binary_layout T, class OPTION_FLAGS>
+struct builder<T, dont_enforce_layout, OPTION_FLAGS>
+{
+    using type = sparrow::string_array;
+
+    template<class U>
+    static type create(U && t)
+    {
+        auto flat_list_view = std::ranges::views::join(ensure_value_range(t));
+        u8_buffer<char> data_buffer(flat_list_view);
+
+        auto sizes = t | std::views::transform([](const auto& l){ 
+            return get_size_save(l);
+        });
+ 
+        return type(
+            std::move(data_buffer),
+            type::offset_from_sizes(sizes),
+            where_null(t)
+        );
+    }
+};
+
+template< translate_to_union_layout T, class OPTION_FLAGS>
+struct builder<T, dont_enforce_layout, OPTION_FLAGS>
+{
+    using type = sparrow::sparse_union_array; // TODO use options to select between sparse and dense
+    using variant_type = std::ranges::range_value_t<T>;
+    static constexpr std::size_t variant_size = std::variant_size_v<variant_type>;
+
+
+    template<class U>
+    static type create(U && t) requires(std::is_same_v<type, sparrow::sparse_union_array>)
+    {
+        std::vector<array> detyped_children(variant_size);
+        for_each_index<variant_size>([&](auto i)
+        {
+            using type_at_index = std::variant_alternative_t<decltype(i)::value, variant_type>;
+            auto type_i_col = t | std::views::transform([](const auto& variant)
+            {   
+                return variant.index() == decltype(i)::value ? 
+                    std::get<type_at_index>(variant) : type_at_index{};
+            });
+
+            using layout_policy_type = layout_flag_t<type_at_index>;
+            detyped_children[decltype(i)::value] = array(build_impl<layout_policy_type>(type_i_col,  OPTION_FLAGS{}));
+        });
+
+        // type-ids
+        auto type_id_range = t | std::views::transform([](const auto& v){
+            return static_cast<std::uint8_t>(v.index());
+        });
+        sparrow::u8_buffer<std::uint8_t> type_id_buffer(type_id_range);
+
+        return type(
+            std::move(detyped_children),
+            std::move(type_id_buffer)
+        );
+    }
+};
+
+template< class T, class OPTION_FLAGS>
+struct builder<T, enforce_dict_encoded_layout,OPTION_FLAGS>
+{
+    using key_type = std::uint64_t;
+    using type = sparrow::dictionary_encoded_array<key_type>;
+    // keep the nulls
+    using raw_range_value_type = std::ranges::range_value_t<T>;
+
+    template<class U>
+    static type create(U && t)
+    {   
+        key_type key = 0;
+        std::map<raw_range_value_type, key_type, nested_less<raw_range_value_type>> value_map;
+        std::vector<raw_range_value_type> values;
+        std::vector<key_type> keys;
```

Comment:
Can you reserve (maybe at the possible maximum), then shrink

Reply:
sure, but I think we can skip the shrink